src/expr/node_traversal.cpp

```diff
@@ -0,0 +1,160 @@
+/*********************                                                        */
+/*! \file node_traversal.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Alex Ozdemir
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Iterators for traversing nodes.
+ **/
+
+#include "node_traversal.h"
+
+namespace CVC4 {
+
+NodeDfsIterator::NodeDfsIterator(TNode n, bool postorder)
+    : d_stack{n},
+      d_visited(),
+      d_postorder(postorder),
+      d_initialized(false),
+      d_current(TNode())
+{
+}
+
+NodeDfsIterator::NodeDfsIterator(bool postorder)
+    : d_stack(),
+      d_visited(),
+      d_postorder(postorder),
+      d_initialized(true),
+      d_current(TNode())
+{
+}
+
+NodeDfsIterator& NodeDfsIterator::operator++()
+{
+  // If we were just constructed, advance to first visit
+  initializeIfUninitialized();
```

Comment:
Why do we still have to call `initializeIfUninitialized` here if we later call `advanceToNextVisit` anyways?

Reply:
Yeah. This is a bit ugly, but necessary. Short reason: if we're unintialized then we need to call `advanceToNextVisit` twice.

Why? If we're uninitialized, and `operator++` is called, then we need to (a) go to the first visit, and then (b) go to the next one.

Reply:
OK, can you please add more documentation to make this clear?

Reply:
Sure. I added a bit of explanation inline in 836575c38. Let me know what you think.